sys-utils/lsns.c

```diff
@@ -1031,66 +1121,31 @@ static void add_scols_line(struct lsns *ls, struct libscols_table *table,
 		return;
 	}
 
-	for (i = 0; i < ncolumns; i++) {
-		char *str = NULL;
+	for (i = 0; i < ncolumns; i++)
+		fill_column(ls, ns, proc, line, i);
```

Comment:
Your implementation will fill all the columns _before_ you evaluate the filter. You need to do it _after_ the filter evaluation. See lsblk.

Yyou need something like (pseudo code without return code checks, etc.):
```C
/* create a line */
line = scols_table_new_line( ... );

/* evaluate filter */
if (ls->filter) {
     ...
    /* generate data on demand by filter */
    scols_filter_set_filler_cb( ... );
    ...
    scols_line_apply_filter(....,  &status);
    ...
    if (status == 0) {
         scols_table_remove_line(line);
         return;
    }
}
 
/* fill missing data to the line */
for (i = 0; ln && i < ncolumns; i++) {
           if (scols_line_is_filled(line, i))
                    continue;
           fill_column(ls, ns, proc, line, i);
}       
````

Reply:
Thank you for the explanation. Updated.